File: stml_cpp/tests/test_json.cpp
/// Minimal JSON parser for test verification only.
/// Parses a JSON string into AstNode for comparison with parser output.

#include "test_json.h"
#include <cctype>
#include <stdexcept>
#include <string>

namespace test_json {

class Parser {
public:
    explicit Parser(const std::string& text) : m_text(text), m_pos(0), m_len(static_cast<int>(text.size())) {}

    stml::AstNode parse_value() {
        skip_ws();
        if (m_pos >= m_len) throw std::runtime_error("Unexpected EOF");

        char ch = m_text[m_pos];
        if (ch == '"') return parse_string();
        if (ch == '{') return parse_object();
        if (ch == '[') return parse_array();
        if (ch == 'n') {
            expect("null");
            return stml::AstNode();
        }
        if (ch == 't' || ch == 'f') {
            // booleans → treated as strings in STML
            return parse_string();
        }
        if (ch == '-' || (ch >= '0' && ch <= '9')) {
            return parse_number();
        }
        throw std::runtime_error(std::string("Unexpected char: ") + ch);
    }

private:
    void skip_ws() {
        while (m_pos < m_len) {
            char ch = m_text[m_pos];
            if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r')
                ++m_pos;
            else
                break;
        }
    }

    stml::AstNode parse_string() {
        ++m_pos; // skip opening "
        std::string result;
        while (m_pos < m_len) {
            char ch = m_text[m_pos++];
            if (ch == '"') {
                return stml::AstNode(result);
            }
            if (ch == '\\') {
                if (m_pos >= m_len) throw std::runtime_error("Unexpected EOF in escape");
                char esc = m_text[m_pos++];
                switch (esc) {
                    case '"':  result.push_back('"'); break;
                    case '\\': result.push_back('\\'); break;
                    case '/':  result.push_back('/'); break;
                    case 'n':  result.push_back('\n'); break;
                    case 'r':  result.push_back('\r'); break;
                    case 't':  result.push_back('\t'); break;
                    case 'b':  result.push_back('\b'); break;
                    case 'f':  result.push_back('\f'); break;
                    case 'u': {
                        // Simple unicode escape (only handles BMP)
                        if (m_pos + 4 > m_len) throw std::runtime_error("Unexpected EOF in \\u");
                        std::string hex = m_text.substr(m_pos, 4);
                        m_pos += 4;
                        int codepoint = std::stoi(hex, nullptr, 16);
                        if (codepoint < 0x80) {
                            result.push_back(static_cast<char>(codepoint));
                        } else if (codepoint < 0x800) {
                            result.push_back(static_cast<char>(0xC0 | (codepoint >> 6)));
                            result.push_back(static_cast<char>(0x80 | (codepoint & 0x3F)));
                        } else {
                            result.push_back(static_cast<char>(0xE0 | (codepoint >> 12)));
                            result.push_back(static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F)));
                            result.push_back(static_cast<char>(0x80 | (codepoint & 0x3F)));
                        }
                        break;
                    }
                    default: result.push_back('\\'); result.push_back(esc); break;
                }
            } else {
                result.push_back(ch);
            }
        }
        throw std::runtime_error("Unclosed string");
    }

    stml::AstNode parse_object() {
        ++m_pos; // skip {
        stml::AstMap map;
        skip_ws();
        if (m_pos < m_len && m_text[m_pos] == '}') {
            ++m_pos;
            return stml::AstNode(std::move(map));
        }
        while (true) {
            skip_ws();
            if (m_text[m_pos] != '"') throw std::runtime_error("Expected string key");
            stml::AstNode key_node = parse_string();
            std::string key = *key_node.as_string();
            skip_ws();
            if (m_text[m_pos++] != ':') throw std::runtime_error("Expected ':'");
            skip_ws();
            stml::AstNode val = parse_value();
            map.emplace_back(std::move(key), std::move(val));
            skip_ws();
            if (m_pos < m_len && m_text[m_pos] == ',') {
                ++m_pos;
                continue;
            }
            if (m_text[m_pos] == '}') {
                ++m_pos;
                break;
            }
            throw std::runtime_error("Expected ',' or '}' in object");
        }
        return stml::AstNode(std::move(map));
    }

    stml::AstNode parse_array() {
        ++m_pos; // skip [
        stml::AstList list;
        skip_ws();
        if (m_pos < m_len && m_text[m_pos] == ']') {
            ++m_pos;
            return stml::AstNode(std::move(list));
        }
        while (true) {
            skip_ws();
            list.push_back(parse_value());
            skip_ws();
            if (m_pos < m_len && m_text[m_pos] == ',') {
                ++m_pos;
                continue;
            }
            if (m_text[m_pos] == ']') {
                ++m_pos;
                break;
            }
            throw std::runtime_error("Expected ',' or ']' in array");
        }
        return stml::AstNode(std::move(list));
    }

    stml::AstNode parse_number() {
        int start = m_pos;
        if (m_text[m_pos] == '-') ++m_pos;
        while (m_pos < m_len && std::isdigit(m_text[m_pos])) ++m_pos;
        if (m_pos < m_len && m_text[m_pos] == '.') {
            ++m_pos;
            while (m_pos < m_len && std::isdigit(m_text[m_pos])) ++m_pos;
        }
        if (m_pos < m_len && (m_text[m_pos] == 'e' || m_text[m_pos] == 'E')) {
            ++m_pos;
            if (m_pos < m_len && (m_text[m_pos] == '+' || m_text[m_pos] == '-')) ++m_pos;
            while (m_pos < m_len && std::isdigit(m_text[m_pos])) ++m_pos;
        }
        // All STML scalars are strings, so return as string
        return stml::AstNode(m_text.substr(start, m_pos - start));
    }

    void expect(const char* word) {
        while (*word) {
            if (m_pos >= m_len || m_text[m_pos++] != *word++)
                throw std::runtime_error("Expected " + std::string(word));
        }
    }

    std::string m_text;
    int m_pos;
    int m_len;
};

stml::AstNode parse(const std::string& json_text) {
    Parser p(json_text);
    return p.parse_value();
}

stml::AstNode load(const std::string& filepath) {
    std::ifstream file(filepath);
    if (!file.is_open()) {
        throw std::runtime_error("Cannot open file: " + filepath);
    }
    std::string text((std::istreambuf_iterator<char>(file)),
                      std::istreambuf_iterator<char>());
    return parse(text);
}

} // namespace test_json

File: stml_java/src/test/java/stml/TestJson.java
package stml;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public class TestJson {
    private String text;
    private int pos;
    private int len;

    public TestJson(String text) {
        this.text = text;
        this.pos = 0;
        this.len = text.length();
    }

    public AstNode parseValue() {
        skipWs();
        if (pos >= len) throw new RuntimeException("Unexpected EOF");
        char ch = text.charAt(pos);
        if (ch == '"') return parseString();
        if (ch == '{') return parseObject();
        if (ch == '[') return parseArray();
        if (ch == 'n') { expect("null"); return new AstNode(); }
        if (ch == 't' || ch == 'f') return parseString();
        if (ch == '-' || (ch >= '0' && ch <= '9')) return parseNumber();
        throw new RuntimeException("Unexpected char: " + ch);
    }

    private void skipWs() {
        while (pos < len) {
            char ch = text.charAt(pos);
            if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r') pos++;
            else break;
        }
    }

    private AstNode parseString() {
        pos++; // skip opening "
        StringBuilder result = new StringBuilder();
        while (pos < len) {
            char ch = text.charAt(pos++);
            if (ch == '"') return new AstNode(result.toString());
            if (ch == '\\') {
                if (pos >= len) throw new RuntimeException("Unexpected EOF in escape");
                char esc = text.charAt(pos++);
                switch (esc) {
                    case '"': result.append('"'); break;
                    case '\\': result.append('\\'); break;
                    case '/': result.append('/'); break;
                    case 'n': result.append('\n'); break;
                    case 'r': result.append('\r'); break;
                    case 't': result.append('\t'); break;
                    case 'b': result.append('\b'); break;
                    case 'f': result.append('\f'); break;
                    case 'u': {
                        if (pos + 4 > len) throw new RuntimeException("Unexpected EOF in \\u");
                        String hex = text.substring(pos, pos + 4);
                        pos += 4;
                        int codepoint = Integer.parseInt(hex, 16);
                        result.append((char) codepoint);
                        break;
                    }
                    default: result.append('\\'); result.append(esc); break;
                }
            } else {
                result.append(ch);
            }
        }
        throw new RuntimeException("Unclosed string");
    }

    private AstNode parseObject() {
        pos++; // skip {
        List<AstNode.Pair> map = new ArrayList<>();
        skipWs();
        if (pos < len && text.charAt(pos) == '}') {
            pos++;
            return new AstNode(map, AstNode.Kind.MAP);
        }
        while (true) {
            skipWs();
            if (text.charAt(pos) != '"') throw new RuntimeException("Expected string key");
            AstNode keyNode = parseString();
            String key = keyNode.asString();
            skipWs();
            if (text.charAt(pos++) != ':') throw new RuntimeException("Expected ':'");
            skipWs();
            AstNode val = parseValue();
            map.add(new AstNode.Pair(key, val));
            skipWs();
            if (pos < len && text.charAt(pos) == ',') { pos++; continue; }
            if (text.charAt(pos) == '}') { pos++; break; }
            throw new RuntimeException("Expected ',' or '}' in object");
        }
        return new AstNode(map);
    }

    private AstNode parseArray() {
        pos++; // skip [
        List<AstNode> list = new ArrayList<>();
        skipWs();
        if (pos < len && text.charAt(pos) == ']') {
            pos++;
            return new AstNode(list, AstNode.Kind.LIST);
        }
        while (true) {
            skipWs();
            list.add(parseValue());
            skipWs();
            if (pos < len && text.charAt(pos) == ',') { pos++; continue; }
            if (text.charAt(pos) == ']') { pos++; break; }
            throw new RuntimeException("Expected ',' or ']' in array");
        }
        return new AstNode(list);
    }

    private AstNode parseNumber() {
        int start = pos;
        if (text.charAt(pos) == '-') pos++;
        while (pos < len && Character.isDigit(text.charAt(pos))) pos++;
        if (pos < len && text.charAt(pos) == '.') {
            pos++;
            while (pos < len && Character.isDigit(text.charAt(pos))) pos++;
        }
        if (pos < len && (text.charAt(pos) == 'e' || text.charAt(pos) == 'E')) {
            pos++;
            if (pos < len && (text.charAt(pos) == '+' || text.charAt(pos) == '-')) pos++;
            while (pos < len && Character.isDigit(text.charAt(pos))) pos++;
        }
        return new AstNode(text.substring(start, pos));
    }

    private void expect(String word) {
        for (int i = 0; i < word.length(); i++) {
            if (pos >= len || text.charAt(pos++) != word.charAt(i))
                throw new RuntimeException("Expected " + word);
        }
    }

    public static AstNode parse(String jsonText) {
        return new TestJson(jsonText).parseValue();
    }

    public static AstNode load(String filepath) throws IOException {
        String text = Files.readString(Path.of(filepath));
        return parse(text);
    }
}